src/analysis/fourier.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import measure_theory.l2_space
+import measure_theory.haar_measure
+import geometry.manifold.instances.circle
+
+/-!
+
+# Fourier analysis on the circle
+
+This file contains some first steps towards Fourier series.
+
+## Main definitions
+
+* `haar_circle`, Haar measure on the circle, normalized to have total measure `1`
+* instances `measure_space`, `probability_measure` for the circle with respect to this measure
+* for `n : ℤ`, `fourier n` is the monomial `λ z, z ^ n`, bundled as a continuous map from `circle`
+  to `ℂ`
+
+## Main statements
+
+The theorem `orthonormal_fourier` states that the functions `fourier n`, when sent via
+`continuous_map.to_Lp` to the L^2 space on the circle, form an orthonormal set.
+
+## TODO
+
+* Show that `submodule.span fourier` is dense in `C(circle, ℂ)`, i.e. that its
+  `submodule.topological_closure` is `⊤`.  This follows from the Stone-Weierstrass theorem after
+  checking that it is a subalgebra, closed under conjugation, and separates points.
+* Show that the image of `submodule.span fourier` under `continuous_map.to_Lp` is dense in the
+  `L^2` space on the circle, and thus that the functions `fourier` form a Hilbert basis.  This
+  follows from the previous result using general theory on approximation by continuous functions.
+
+-/
+
+noncomputable theory
+open topological_space continuous_map measure_theory measure_theory.measure
+
+local attribute [instance] fact_one_le_two_ennreal
+
+section haar_circle
+/-! We make the circle into a measure space, using the Haar measure normalized to have total
+measure 1. -/
+
+instance : measurable_space circle := borel circle
+instance : borel_space circle := ⟨rfl⟩
+
+/-- Haar measure on the circle, normalized to have total measure 1. -/
+def haar_circle : measure circle := haar_measure positive_compacts_univ
+
+instance : probability_measure haar_circle := ⟨haar_measure_self⟩
+
+instance : measure_space circle :=
+{ volume := haar_circle,
+  .. circle.measurable_space }
+
+end haar_circle
+
+section fourier
+
+/-- The family of monomials `λ z, z ^ n`, parametrized by `n : ℤ` and considered as bundled
+continuous maps from `circle` to `ℂ`. -/
+def fourier (n : ℤ) : C(circle, ℂ) :=
+{ to_fun := λ z, z ^ n,
+  continuous_to_fun := continuous_subtype_coe.fpow nonzero_of_mem_circle n }
+
+@[simp] lemma fourier_apply {n : ℤ} {z : circle} : fourier n z = z ^ n := rfl
+
+@[simp] lemma fourier_zero {z : circle} : fourier 0 z = 1 := rfl
+
+lemma conj_fourier {n : ℤ} {z : circle} : complex.conj (fourier n z) = fourier (-n) z :=
```

Comment:
```suggestion
@[simp] lemma conj_fourier {n : ℤ} {z : circle} : complex.conj (fourier n z) = fourier (-n) z :=
```

Reply:
@semorrison My impression is that making `conj_fourier` (here) and `mul_fourier` simp-lemmas will conflict with making `fourier_apply` a simp-lemma, for example, see below in the proof of `orthonormal_fourier` where I want to simp using these lemmas, and have to explicitly remove `fourier_apply` from the simp-set.

You know much more about the simplifier than I do; do you agree?  In general, this issue of simplifying via function application as opposed to simplifying some constructed function itself must have arisen before.

Reply:
Sorry I didn't get back to this. Yes, I was wrong and if you want `fourier_apply` to be a simp lemma then you can't have / don't need these as well.

Reply:
I noticed that if you reverse the equality in `conj_fourier` and `mul_fourier`, they at least become valid as simp-lemmas, so I have done that (with appropriate renaming).  That means that in the lemma `orthonormal_fourier` which I had called out, they get used backwards.  What do you think?  I can switch it back if that would be better.